src/support/generation.c

```diff
@@ -127,14 +127,21 @@ __wt_gen_drain(WT_SESSION_IMPL *session, int which, uint64_t generation)
      * session count. That way, no matter what sessions come or go, we'll check the slots for all of
      * the sessions that could have been active when we started our check.
      */
-    WT_C_MEMMODEL_ATOMIC_LOAD(session_cnt, &conn->session_cnt, WT_ATOMIC_SEQ_CST);
+    WT_C_MEMMODEL_ATOMIC_LOAD(session_cnt, &conn->session_cnt, WT_ATOMIC_ACQUIRE);
     for (minutes = 0, pause_cnt = 0, s = conn->sessions, i = 0; i < session_cnt; ++s, ++i) {
         if (!s->active)
             continue;
 
         for (;;) {
-            /* Ensure we only read the value once. */
-            WT_C_MEMMODEL_ATOMIC_LOAD(v, &s->generations[which], WT_ATOMIC_SEQ_CST);
+            /*
+             * Ensure we only read the value once.
+             *
+             * The requirement is that we should see all the sessions that enter the generation
+             * before the global generation we read. This is already guranteed when we read the
+             * global generation using sequential consistency. Therefore, here we can get away using
+             * a relaxed read.
+             */
+            WT_C_MEMMODEL_ATOMIC_LOAD(v, &s->generations[which], WT_ATOMIC_RELAXED);
```

Comment:
I don't think this is correct, this could read an older version of `s->generations[which]`, in the running examples we have it could read it as 0 even though it's been set to 5.

This [CPPMem](http://svr-pes20-cppmem.cl.cam.ac.uk/cppmem/) snippet highlights the problem, the first thread runs `__wt_gen_next_drain` and the second `__wt_session_gen_enter`:
```
int main() {
  atomic_int s = 0; // session generation
  atomic_int c = 5; // connection generation
  {{{ 
    { // __wt_gen_next_drain
     c.load(); c.store(6);             // fetch_and_add, moved to session 6
     s.load(mo_relaxed).readsvalue(0); // faulty relaxed load
                                       // we assume nobody in session 5
    }
  ||| 
    { // __wt_session_gen_enter
     s.store(c.load().readsvalue(5)); 
     c.load().readsvalue(5);           // succesfully entered session 5
    }
  }}};
  return 0; }
```

Reply:
Thanks @dgomezferro. That makes sense to me. The mistake I made is that I forgot this operation s.store(c.load().readsvalue(5)) is not atomic.

Reply:
@dgomezferro Sorry, the real issue is not what I described above. The example in the [cppreference](https://en.cppreference.com/w/cpp/atomic/memory_order#Sequentially-consistent_ordering) website is clear. The following example is allowed to produce the result r1 == 1 && r2 == 3 && r3 == 0

```
// Thread 1:
x.store(1, std::memory_order_seq_cst); // A
y.store(1, std::memory_order_release); // B
// Thread 2:
r1 = y.fetch_add(1, std::memory_order_seq_cst); // C
r2 = y.load(std::memory_order_relaxed); // D
// Thread 3:
y.store(3, std::memory_order_seq_cst); // E
r3 = x.load(std::memory_order_seq_cst); // F
```

Note that:

1) as soon as atomic operations that are not tagged memory_order_seq_cst enter the picture, the sequential consistency guarantee for the program is lost
2) in many cases, memory_order_seq_cst atomic operations are reorderable with respect to other atomic operations performed by the same thread

in many cases, memory_order_seq_cst atomic operations are reorderable with respect to other atomic operations performed by the same thread

This is a surprise to me but it makes sense. It is not reoderable only for read-modify-write operations as it has both acquire and release semantics. But for a load or store sequential consistency operations, they are still reorderable so we can see the behaviour in the example.

Reply:
You're right, I couldn't find the RMW operation at first on cppmen, but it has CAS functions, this is the correct model I think:

```
int main() {
  atomic_int s = 0; // session generation
  atomic_int c = 5; // connection generation
  {{{ 
    { // __wt_gen_next_drain
     cas_strong(c, 5, 6);              // fetch_and_add, moved to generation 6
     s.load(mo_relaxed).readsvalue(0); // problematic relaxed load
                                       // we assume nobody in generation 5
    }
  ||| 
    { // __wt_session_gen_enter
     s.store(c.load().readsvalue(5)); 
     c.load().readsvalue(5);           // succesfully entered generation 5
    }
  }}};
  return 0; }
```

It still shows the reordering is possible.

Reply:
I'm not sure then, cannot explain how this is possible based on my understanding. cas_strong(c, 5, 6); must be sequenced after c.load().readsvalue(5); in the total ordering. That means in the total ordering the first thread should be able to see this write s.store(c.load().readsvalue(5)). @dgomezferro Do you see any issue in my reasoning?

Reply:
>  cas_strong(c, 5, 6); must be sequenced after c.load().readsvalue(5); in the total ordering.

And it is, you can follow the orange `sc` arrows for the sequentially consistent ordering:
<img width="387" alt="image" src="https://github.com/wiredtiger/wiredtiger/assets/751899/9e09cced-7ba6-41a0-b263-54abe3e79d0d">

> That means in the total ordering the first thread should be able to see this write s.store(c.load().readsvalue(5))
No, the relaxed load doesn't have to honor the total order established by the seq_cst operations, it does its own thing and reads from the past. 

This is what this cppreference paragraph is warning about:
> as soon as atomic operations that are not tagged memory_order_seq_cst enter the picture, the sequential consistency guarantee for the program is lost.

Reply:
```
No, the relaxed load doesn't have to honor the total order established by the seq_cst operations, it does its own thing and reads from the past. In our case it reads from the initial value (follow the rf arrow).
```
I also tried using an acquire read but it gives me the same result. It feels magic to me the visibility of sequential consistency is not transitive and it is possible to read from the past even the read cannot be reordered before the cas.

Reply:
I agree it's pretty counter intuitive, I also thought this optimization was possible while reviewing your PR but couldn't make it work with a litmus test, then kind of understood why that wasn't possible.

I think the explanation is that the total order only applies to seq_cst operations, and this total order doesn't create happens-before or synchronizes-with relationships. Non-seq_cst operations are only ordered with respect to operations on other threads through synchronizes-with relationships (ignoring release-consume ordering here...), and the RMW operation doesn't synchronize-with the thread entering the generation.